### PR TITLE
Lockless FileExistsMap and position mapping

### PR DESCRIPTION
**Pull request by pepeiborra**
5th instalment from #2357 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2442"></a>

### CODE REVIEW

File: ghcide/src/Development/IDE/Core/FileExists.hs [hunk 1]
```diff
@@ -75,32 +74,34 @@ fast path by a check that the path also matches our watching patterns.
 -- | A map for tracking the file existence.
 -- If a path maps to 'True' then it exists; if it maps to 'False' then it doesn't exist'; and
 -- if it's not in the map then we don't know.
-type FileExistsMap = (HashMap NormalizedFilePath Bool)
+type FileExistsMap = STM.Map NormalizedFilePath Bool
 
 -- | A wrapper around a mutable 'FileExistsState'
-newtype FileExistsMapVar = FileExistsMapVar (Var FileExistsMap)
+newtype FileExistsMapVar = FileExistsMapVar FileExistsMap
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/FileExists.hs:80]: is this newtype still useful? I guess for the `IsIdeGlobal` instance?

File: ghcide/src/Development/IDE/Core/FileExists.hs [hunk 2]
```diff
@@ -75,32 +74,34 @@ fast path by a check that the path also matches our watching patterns.
...
   FileExistsMapVar v <- getIdeGlobalAction
-  liftIO $ readVar v
+  return v
 
 -- | Modify the global store of file exists.
 modifyFileExists :: IdeState -> [(NormalizedFilePath, FileChangeType)] -> IO ()
 modifyFileExists state changes = do
   FileExistsMapVar var <- getIdeGlobalState state
-  changesMap           <- evaluate $ HashMap.fromList changes
   -- Masked to ensure that the previous values are flushed together with the map update
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/FileExists.hs:94]: The body of the do block is now two... STM actions! so can we just join them together into one transaction and drop this?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 3]
```diff
@@ -346,7 +345,7 @@ lastValueIO s@ShakeExtras{positionMapping,persistentKeys,state} k file = do
                 return Nothing
             Just (v,del,ver) -> do
                 void $ atomically $ STM.focus (Focus.alter (alterValue $ Stale (Just del) ver (toDyn v))) (toKey k file) state
-                return $ Just (v,addDelta del $ mappingForVersion allMappings file ver)
+                atomically $ Just . (v,) . addDelta del <$> mappingForVersion positionMapping file ver
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:348]: Merge transaction with the one above?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 4]
```diff
@@ -359,8 +358,10 @@ lastValueIO s@ShakeExtras{positionMapping,persistentKeys,state} k file = do
     atomically (STM.lookup (toKey k file) state) >>= \case
       Nothing -> readPersistent
       Just (ValueWithDiagnostics v _) -> case v of
-        Succeeded ver (fromDynamic -> Just v) -> pure (Just (v, mappingForVersion allMappings file ver))
-        Stale del ver (fromDynamic -> Just v) -> pure (Just (v, maybe id addDelta del $ mappingForVersion allMappings file ver))
+        Succeeded ver (fromDynamic -> Just v) ->
+            atomically $ Just . (v,) <$> mappingForVersion positionMapping file ver
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:362]: Just use one transaction including these and the stuff on line 358? Hm, maybe `readPersistent` gets in the way...

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 5]
```diff
@@ -1223,17 +1223,16 @@ getAllDiagnostics =
     fmap (concatMap (\(k,v) -> map (fromUri k,ShowDiag,) $ getDiagnosticsFromStore v)) . ListT.toList . STM.listT
 
 updatePositionMapping :: IdeState -> VersionedTextDocumentIdentifier -> List TextDocumentContentChangeEvent -> IO ()
-updatePositionMapping IdeState{shakeExtras = ShakeExtras{positionMapping}} VersionedTextDocumentIdentifier{..} (List changes) = do
-    modifyVar_ positionMapping $ \allMappings -> do
-        let uri = toNormalizedUri _uri
-        let mappingForUri = HMap.lookupDefault Map.empty uri allMappings
-        let (_, updatedMapping) =
+updatePositionMapping IdeState{shakeExtras = ShakeExtras{positionMapping}} VersionedTextDocumentIdentifier{..} (List changes) =
+    atomically $ STM.focus (Focus.alter f) uri positionMapping
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:1227]: push to call sites?

File: ghcide/src/Development/IDE/Core/Shake.hs [hunk 6]
```diff
@@ -328,7 +328,6 @@ getIdeOptionsIO ide = do
 -- for the version of that value.
 lastValueIO :: IdeRule k v => ShakeExtras -> k -> NormalizedFilePath -> IO (Maybe (v, PositionMapping))
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/Shake.hs:329]: Interestingly, I think this entire function is close to being in STM. Maybe just `persistentKeys` blocking it. Not sure if that's desirable.

**Review by michaelpj — APPROVED** [review]

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 6 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs:329]: I think this one is firmly in `IO`

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 5 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs]: Done

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 4 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs:362]: It does

File: ghcide/src/Development/IDE/Core/Shake.hs (hunk 3 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Shake.hs]: Done

File: ghcide/src/Development/IDE/Core/FileExists.hs (hunk 2 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/FileExists.hs:95]: We can. The `mask_` around them is enough to mostly ensure correctness, but a single transaction is definitely correct'er.

File: ghcide/src/Development/IDE/Core/FileExists.hs (hunk 1 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/FileExists.hs:81]: Yes, still needed

**Pull request merged**